app_dart/test/request_handlers/scheduler/batch_backfiller_test.dart

```diff
@@ -40,325 +40,890 @@ void main() {
+    expect(pubsub.messages.length, 1);
+
+    expect(
+      firestore,
+      existsInStorage(fs.Task.metadata, [
+        isTask.hasStatus(fs.Task.statusSucceeded).hasCommitSha('1'),
+        isTask.hasStatus(fs.Task.statusSucceeded).hasCommitSha('2'),
+        isTask.hasStatus(fs.Task.statusInProgress).hasCommitSha('3'),
+      ]),
+    );
+  });
+
+  test('updates task as in-progress after backfilling', () async {
+    final oldestGray = <Task>[
+      generateTask(1, name: 'Linux_android A', status: Task.statusSucceeded),
+      generateTask(2, name: 'Linux_android A', status: Task.statusSucceeded),
+      generateTask(3, name: 'Linux_android A', status: Task.statusNew),
+    ];
+    db.addOnQuery<Task>((Iterable<Task> results) => oldestGray);
+    final task = oldestGray[2];
+    expect(db.values.length, 0);
+    expect(task.status, Task.statusNew);
+
+    firestore.putDocument(
+      generateFirestoreTask(
+        1,
+        name: 'Linux_android A',
+        status: fs.Task.statusSucceeded,
+        commitSha: '1',
+      ),
+    );
+    firestore.putDocument(
+      generateFirestoreTask(
+        2,
+        name: 'Linux_android A',
+        status: fs.Task.statusSucceeded,
+        commitSha: '2',
+      ),
+    );
+    firestore.putDocument(
+      generateFirestoreTask(
+        3,
+        name: 'Linux_android A',
+        status: fs.Task.statusNew,
+        commitSha: '3',
+      ),
+    );
+
+    await tester.get(handler);
+    expect(db.values.length, 1);
+    expect(task.status, Task.statusInProgress);
+
+    expect(
+      firestore,
+      existsInStorage(fs.Task.metadata, [
+        isTask.hasStatus(fs.Task.statusSucceeded),
+        isTask.hasStatus(fs.Task.statusSucceeded),
+        isTask.hasStatus(fs.Task.statusInProgress),
+      ]),
+    );
+  });
+
+  test('skip scheduling builds if datastore commit fails', () async {
+    db.commitException = true;
+    final oldestGray = <Task>[
+      generateTask(1, name: 'Linux_android A', status: Task.statusSucceeded),
+      generateTask(2, name: 'Linux_android A', status: Task.statusSucceeded),
+      generateTask(3, name: 'Linux_android A', status: Task.statusNew),
+    ];
+    db.addOnQuery<Task>((Iterable<Task> results) => oldestGray);
+    expect(db.values.length, 0);
+
+    firestore.putDocument(
+      generateFirestoreTask(
+        1,
+        name: 'Linux_android A',
+        status: fs.Task.statusSucceeded,
+        commitSha: '1',
+      ),
+    );
+    firestore.putDocument(
+      generateFirestoreTask(
+        2,
+        name: 'Linux_android A',
+        status: fs.Task.statusSucceeded,
+        commitSha: '2',
+      ),
+    );
+    firestore.putDocument(
+      generateFirestoreTask(
+        3,
+        name: 'Linux_android A',
+        status: fs.Task.statusNew,
+        commitSha: '3',
+      ),
+    );
+
+    await tester.get(handler);
+    expect(db.values.length, 0);
+    expect(pubsub.messages.length, 0);
+
+    expect(
```

Comment:
Had to re-read the test description before it clicked why these would be the same.